### PR TITLE
fix: handle Gemini 2.5 JSON mode + function calling conflict

### LINKED ISSUE #3964: [Bug] Response mime type not accepted 400 error
### Description

DEBUG Files added: 1
ERROR    Error from Gemini API: 400 INVALID_ARGUMENT. {'error': {'code': 400, 'message': 'Unable to submit request because  
         it has a mimeType parameter with value application/json, which is not supported. Update the mimeType and try again.
         Learn more: https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/gemini', 'status':
         'INVALID_ARGUMENT'}}
WARNING  Attempt 1/1 failed: <Response [400 Bad Request]>
ERROR    Failed after 1 attempts. Last error using Gemini(gemini-2.0-flash)
ERROR    ❌ Step 1 failed: <Response [400 Bad Request]>

my code snippet was, - 
 content_generation_schema = {
            "type": "object",
            "properties": {
                "section_title": {
                    "type": "string",
                    "description": "The exact subtopic title"
                },
                "section_parts": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "properties": {
                            "part_type": {
                                "type": "string",
                                "enum": ["Introduction", "Definition", "Key Concepts", "Practical Applications"]
                            },
                            "content": {
                                "type": "string",
                                "description": "Educational content for this part with proper formatting"
                            },
                            "requires_diagram": {
                                "type": "string",
                                "description": "Type of diagram needed to enhance this section, if applicable",
                                "enum": ["mindmap", "flowchart", "erdiagram", "timeline", "xychart", "piechart", "userjourney", "classdiagram", "none"]
                            },
                            "diagram_desc": {
                                "type": "string",
                                "description": "Detailed description of what should be in the diagram, used as query to generate the diagram"
                            }
                        },
                        "required": ["part_type", "content", "requires_diagram", "diagram_desc"]
                    },
                    "minItems": 4,
                    "maxItems": 4
                }
            },
            "required": ["section_title", "section_parts"]
        }
        
        return Agent(
            name="ML Content Generator",
            role="Expert educational content writer",
            model=Gemini(
                id="gemini-2.0-flash",
                temperature=0.4,
                generation_config={
                    # "response_mime_type": "application/json",
                    "response_schema": content_generation_schema
                }
            ),
            instructions=[
                "You are an expert educational content writer for Machine Learning.",
                "Generate detailed educational notes based on subtopic descriptions.",
                "Structure content into Introduction, Definition, Key Concepts, and Practical Applications.",
                "Include diagram specifications where appropriate.",
                "Use proper educational formatting and cross-references to other modules.",
                "Return structured JSON with section_title, section_parts array containing part_type, content, requires_diagram, and diagram_desc fields."
            ],
            debug_mode=True,
            show_tool_calls=False
        )

### Steps to Reproduce

 content_generation_schema = {
            "type": "object",
            "properties": {
                "section_title": {
                    "type": "string",
                    "description": "The exact subtopic title"
                },
                "section_parts": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "properties": {
                            "part_type": {
                                "type": "string",
                                "enum": ["Introduction", "Definition", "Key Concepts", "Practical Applications"]
                            },
                            "content": {
                                "type": "string",
                                "description": "Educational content for this part with proper formatting"
                            },
                            "requires_diagram": {
                                "type": "string",
                                "description": "Type of diagram needed to enhance this section, if applicable",
                                "enum": ["mindmap", "flowchart", "erdiagram", "timeline", "xychart", "piechart", "userjourney", "classdiagram", "none"]
                            },
                            "diagram_desc": {
                                "type": "string",
                                "description": "Detailed description of what should be in the diagram, used as query to generate the diagram"
                            }
                        },
                        "required": ["part_type", "content", "requires_diagram", "diagram_desc"]
                    },
                    "minItems": 4,
                    "maxItems": 4
                }
            },
            "required": ["section_title", "section_parts"]
        }
        
        return Agent(
            name="ML Content Generator",
            role="Expert educational content writer",
            model=Gemini(
                id="gemini-2.0-flash",
                temperature=0.4,
                generation_config={
                    # "response_mime_type": "application/json",
                    "response_schema": content_generation_schema
                }
            ),
            instructions=[
                "You are an expert educational content writer for Machine Learning.",
                "Generate detailed educational notes based on subtopic descriptions.",
                "Structure content into Introduction, Definition, Key Concepts, and Practical Applications.",
                "Include diagram specifications where appropriate.",
                "Use proper educational formatting and cross-references to other modules.",
                "Return structured JSON with section_title, section_parts array containing part_type, content, requires_diagram, and diagram_desc fields."
            ],
            debug_mode=True,
            show_tool_calls=False
        )

### Agent Configuration (if applicable)

_No response_

### Expected Behavior

response in json format with the given response schema

### Actual Behavior

DEBUG Files added: 1
ERROR    Error from Gemini API: 400 INVALID_ARGUMENT. {'error': {'code': 400, 'message': 'Unable to submit request because  
         it has a mimeType parameter with value application/json, which is not supported. Update the mimeType and try again.
         Learn more: https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/gemini', 'status':
         'INVALID_ARGUMENT'}}
WARNING  Attempt 1/1 failed: <Response [400 Bad Request]>
ERROR    Failed after 1 attempts. Last error using Gemini(gemini-2.0-flash)
ERROR    ❌ Step 1 failed: <Response [400 Bad Request]>

### Screenshots or Logs (if applicable)

_No response_

### Environment

```markdown
Windows,  agno 1.7.5, gemini 2.5/2.0 flash
```

### Possible Solutions (optional)

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Mustafa-Esoofally**
## Summary

Gemini 2.5 models (flash and pro) reject API requests that combine `response_mime_type='application/json'` with function declarations, returning `400 INVALID_ARGUMENT`. This breaks any agent that uses `output_schema` together with tools — including the implicit `search_knowledge_base` tool registered when `knowledge` is set.

Closes #3964, also addresses #3292.

### The Problem

When a user creates an agent like:

```python
agent = Agent(
    model=Gemini(id="gemini-2.5-flash"),
    output_schema=MovieReview,
    tools=[get_movie_year],
)
```

Agno builds a `GenerateContentConfig` with both `response_mime_type="application/json"` + `response_schema` AND `tools=[Tool(function_declarations=[...])]`. Gemini 2.5 uses strict token-level constrained decoding for JSON output — every generated token is forced to match the schema grammar. Function call tokens can't satisfy JSON schema constraints, so the API rejects the combination upfront.

This only affects Gemini 2.5. Gemini 2.0 uses loose "best effort" JSON mode (no conflict), and [Gemini 3.x implements multi-mode decoding](https://ai.google.dev/gemini-api/docs/gemini-3) that separates tool-calling and JSON output phases (no conflict).

### How Users Hit This Bug

There are **three paths** to the conflict, and only one is obvious:

**Path 1 — Explicit tools + output_schema (obvious):**
```python
agent = Agent(
    model=Gemini(id="gemini-2.5-flash"),
    output_schema=MovieReview,
    tools=[get_movie_year],       # user explicitly adds tools
)
```

**Path 2 — Knowledge + output_schema (invisible, most common):**
```python
agent = Agent(
    model=Gemini(id="gemini-2.5-flash"),
    knowledge=my_knowledge_base,  # this is ALL the user did
    output_schema=ResearchSummary,
)
# User never wrote tools=[...] but search_knowledge=True is the DEFAULT
# → Agno auto-registers search_knowledge_base as a function tool at _tools.py:176
# → function_declarations end up in the config → 400 INVALID_ARGUMENT
```

This is the exact scenario from issue #3964. The user sets `knowledge` and `output_schema` — both completely reasonable on their own. They have no idea a function tool was registered behind the scenes. The only workaround was `search_knowledge=False`, which defeats the purpose of having a knowledge base.

**Path 3 — Other auto-registered tools:**
Same pattern with `read_chat_history=True` or other agent features that silently register function tools.

### Error Swallowing

When the 400 error occurs, Agno doesn't raise an exception — it returns the error JSON string as `response.content`. The user expects a `MovieReview` Pydantic model but gets `'{"error": {"code": 400, ...}}'` as a string, leading to confusing downstream `AttributeError`s.

### Previous Fix Attempts

This bug has existed since Gemini 2.5 launched. There was a partial attempt to address it:

- **Apr 2025, PR #2655** — @dirkbrnd added a warning log and moved `response_mime_type` after tools. But the guard had a typo: `config.get("tool")` instead of `config.get("tools")` — so it never actually fired. And even if it had, it only warned without preventing the conflict.
- **May 2025, PR #2631** — Memory 2.0 refactor removed the warning entirely during restructuring.
- **Sep 2025, PR #4647** — Rewrote the response_format block with `prepare_response_schema`. No conflict guard.

Since May 2025, there has been zero protection against this conflict.

### The Fix

A proactive guard in `Gemini.get_request_params()` — the single chokepoint all 4 invoke variants (sync, async, stream, non-stream) pass through:

1. **Detect the conflict:** Is this a Gemini 2.5 model AND does config have `response_mime_type='application/json'` AND does config contain function declarations?
2. **Strip JSON mode:** Remove `response_mime_type` and `response_schema` from the config so the API accepts the request
3. **Inject JSON instructions:** Append `get_json_output_prompt(schema)` to `system_instruction` — the same utility the agent layer uses for non-native JSON models — so the model still knows to produce JSON
4. **Existing parser handles the rest:** `parse_response_model_str()` already handles prompt-based JSON responses with 4 fallback strategies

### Request Flow (Before → After)

**Before (broken):**
```
Agent.run()
  → _build_system_message(): supports_native_structured_outputs=True
    → SKIPS adding JSON instructions to system prompt (trusts native mode)
  → get_request_params() builds config with BOTH json mode + tools
  → Gemini API → 400 INVALID_ARGUMENT
  → Error returned as string content instead of raising
  → User gets error string, not their Pydantic model
```

**After (fixed):**
```
Agent.run()
  → _build_system_message(): same as before (no change to agent layer)
  → get_request_params() builds config with BOTH json mode + tools
  → Guard detects: 2.5 model + JSON mime + function declarations
  → Strips response_mime_type and response_schema from config
  → Injects get_json_output_prompt(schema) into system_instruction
  → Gemini API → succeeds (tools work, model follows JSON prompt)
  → parse_response_model_str() parses text → Pydantic model
  → User gets their MovieReview object
```

### Scope

The guard is narrowly scoped:
- **Only Gemini 2.5 family** — 2.0 and 3.x work natively, not degraded
- **Only function declarations** — built-in tools (Google Search, URL Context, File Search) don't conflict with JSON mode
- **Both config paths** — handles `response_format` from agent layer AND manual `generation_config` from users

### Compatibility Matrix

| Model | JSON + Function Tools | Guard Fires? |
|-------|----------------------|--------------|
| `gemini-2.0-flash` | Works natively | No |
| `gemini-2.5-flash` | API rejects | Yes → prompt fallback |
| `gemini-2.5-pro-preview-*` | API rejects | Yes → prompt fallback |
| `gemini-3-flash-preview` | [Works natively](https://ai.google.dev/gemini-api/docs/gemini-3) | No |
| `gemini-3-pro-preview` | [Works natively](https://ai.google.dev/gemini-api/docs/gemini-3) | No |
| `gemini-3.1-pro-preview` | [Works natively](https://ai.google.dev/gemini-api/docs/gemini-3) | No |

## Type of change

- [x] Bug fix

---

## Checklist

- [x] Code complies with style guidelines
- [x] Ran format/validation scripts (`./scripts/format.sh` and `./scripts/validate.sh`)
- [x] Self-review completed
- [x] Documentation updated (comments, docstrings)
- [x] Examples and guides: Cookbook added (`structured_output_with_tools.py`)
- [x] Tested in clean environment
- [x] Tests added/updated (8 new unit tests)

---

### Files Changed

- **`libs/agno/agno/models/google/gemini.py`** — Added `_is_gemini_2_5()`, `_has_function_declarations()` helpers and the conflict guard in `get_request_params()`
- **`libs/agno/tests/unit/models/google/test_gemini.py`** — 8 new tests covering guard activation, non-activation, system instruction preservation, and manual config handling
- **`cookbook/90_models/google/gemini/structured_output_with_tools.py`** — Cookbook demonstrating `output_schema` + `tools` on Gemini (all 4 variants: sync, streaming, async, async+streaming)

### Tradeoffs

- Prompt-based JSON is less reliable than native constrained decoding — the model might occasionally produce invalid JSON. The existing `parse_response_model_str()` has 4 fallback strategies that handle most edge cases.
- Minor token overhead from injecting the JSON schema prompt into system instructions.
- When Gemini 2.5 eventually reaches end-of-life, this guard becomes dead code. It's isolated enough to remove cleanly.

### Live API Test Results

All 5 reproduction cases pass against the live Gemini API:

| Case | Scenario | Result |
|------|----------|--------|
| 1 | Structured output only (no tools) | PASS |
| 2 | Tools only (no output_schema) | PASS |
| 3 | output_schema + tools (2.5-flash) | PASS (was 400 error) |
| 4 | output_schema + tools (2.5-flash, streaming) | PASS (was 400 error) |
| 5 | output_schema + tools (2.0-flash baseline) | PASS |

### CODE REVIEW

File: libs/agno/agno/models/google/gemini.py [hunk 1]
```diff
@@ -347,6 +360,29 @@ def get_request_params(
...
+
+        if self._is_gemini_2_5() and has_json_mode and has_func_decls:
+            log_warning(
+                "Gemini 2.5 does not support function calling with response_mime_type='application/json'. "
+                "Falling back to prompt-based JSON output."
+            )
+            config.pop("response_mime_type", None)
+            saved_schema = config.pop("response_schema", None)
+
+            json_prompt = get_json_output_prompt(response_format if response_format is not None else saved_schema)  # type: ignore[arg-type]
```
Comment (uzaxirr) [inline libs/agno/agno/models/google/gemini.py:383]: how does the behaviour looks when the user explecitly sets `generation_config={"response_mime_type": "application/json"}` without passing `response_format`, both `response_format` and `saved_schema` will be `None` this might produce a weird prompt

**Review by Mustafa-Esoofally — COMMENTED** [review]
Good fix for a real user-facing bug. The core approach is sound — detecting the Gemini 2.5 JSON+tools conflict in `get_request_params()` and falling back to prompt-based JSON.

I pushed two fixes:

1. **State mutation bug** — `config = self.generation_config` was a reference (not copy) for dict configs. The guard's `pop()` permanently mutated the model's stored config, breaking JSON mode on subsequent calls. Fixed with `dict(self.generation_config)`.

2. **Dict-shaped tools bypass** — `_has_function_declarations()` only checked `isinstance(tool, Tool)`. When `generation_config` is a `GenerateContentConfig`, `model_dump()` converts tools to dicts, bypassing the guard. Added dict check.

Also added tests for state mutation regression and Gemini 3.x non-activation.

**Remaining minor items (acceptable for v1):**
- `_is_gemini_2_5()` won't match `tunedModels/{id}` format (rare edge case)
- When `response_format is None` and `saved_schema` is a google `Schema` object, `get_json_output_prompt()` produces degraded output (manual `generation_config` edge case)
